gemini/evaluation/enhancing_quality_and_explainability_with_rapideval.ipynb

```diff
@@ -0,0 +1,665 @@
+    "\n",
+    "nest_asyncio.apply()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "39ce45a03542"
+   },
+   "source": [
+    "## Definining functions for ranking using evaluations"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "4437b7608c8e"
+   },
+   "source": [
+    "This section defines the various helper functions to perform pairwise and pointwise evaluations, as well as the logic to combine them \n",
+    "to select the best response and return associated quality metrics and explanation.\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "afc4054f1f70"
+   },
+   "source": [
+    "The function below enable Auto SxS comparisons between pairs of responses using the Rapid Evaluation Web API to be used in conjunction with the python max or sorted inbuilt functions  in order to use pairwise comparions to find the best response or get a ranked list of responses respectively.\n",
+    "The SDK could also be used for this purpose once it supports this functionality. The complete list of metrics for other tasks like summarization SxS can be found on the website linked below."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "ac10531efd0c"
+   },
+   "outputs": [],
+   "source": [
+    "def pairwise_greater(instructions, context, project_id, location, baseline, candidate):\n",
+    "    \"\"\"\n",
+    "    Takes Instructions, Context and two different responses.\n",
+    "    Returns the response which best matches the instructionds/Context for the given\n",
+    "    quality metric ( in this case question answering).\n",
+    "    More details on the web API and different quality metrics which this function\n",
+    "    can be extended to can be found on\n",
+    "    https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/evaluation\n",
+    "    \"\"\"\n",
+    "    creds, _ = auth.default(scopes=[\"https://www.googleapis.com/auth/cloud-platform\"])\n",
+    "\n",
+    "    data = {\n",
+    "        \"pairwise_question_answering_quality_input\": {\n",
+    "            \"metric_spec\": {},\n",
+    "            \"instance\": {\n",
+    "                \"prediction\": candidate,\n",
+    "                \"baseline_prediction\": baseline,\n",
+    "                \"instruction\": instructions,\n",
+    "                \"context\": (context),\n",
+    "            },\n",
+    "        }\n",
+    "    }\n",
+    "\n",
+    "    uri = f\"https://{LOCATION}-aiplatform.googleapis.com/v1beta1/projects/{project_id}/locations/{location}:evaluateInstances\"\n",
+    "    result = google_auth_requests.AuthorizedSession(creds).post(uri, json=data)\n",
+    "    result = result.json()[\"pairwiseQuestionAnsweringQualityResult\"]\n",
+    "    choice = baseline if result[\"pairwiseChoice\"] == \"BASELINE\" else candidate\n",
```

Comment:
Can this not be done with the Vertex AI Python SDK? I would highly recommend using that instead or waiting until it's released before publishing this notebook instead of using raw API calls.

Reply:
resolving as discussed in chat